### PR TITLE
conn,device: enable cryptorouting via PeerAwareEndpoint

### PR DESCRIPTION
**Pull request by raggi**
Introduce an optional extension point for Endpoint that enables a path for WireGuard to inform an integration about the peer public key that is associated with an Endpoint.

The API is expected to return either the same or a new Endpoint in response to this function. A future version of this patch could potentially remove the returned Endpoint, but would require larger integrator changes downstream.

This adds a small per-packet cost that could later be removed with a larger refactor of the wireguard-go interface and Tailscale magicsock code, as well as introducing a generic bound for Endpoint in a device & bind instance.

Updates tailscale/corp#20732

### CODE REVIEW

File: conn/conn.go [hunk 1]
```diff
@@ -84,6 +84,18 @@ type Endpoint interface {
...
 
+// PeerAwareEndpoint is an optional Endpoint specialization for integratons that want to
+// know about the outcome of cryptorouting identification.
+//
+// If they receive a packet from a source they had not pre-identified, to learn the
+// identification WireGuard can derive from the session or handshake.
+//
+// If GetPeerEndpoint returns nil, WireGuard will be unable to respond to the peer
+// until a new endpoint is written by a later packet.
+type PeerAwareEndpoint interface {
```
Comment (bradfitz) [inline conn/conn.go:97]: Not going to make this a struct type instead to make the type check cheaper?

The struct can embed a conn.Endpoint and then make GetPeerEndpoint be a func field.

Or did you measure it and find it not worth it?

**Review by bradfitz — APPROVED** [review]

File: conn/conn.go [hunk 2]
```diff
@@ -84,6 +84,18 @@ type Endpoint interface {
 	SrcIP() netip.Addr
 }
 
+// PeerAwareEndpoint is an optional Endpoint specialization for integratons that want to
```
Comment (jwhited) [inline conn/conn.go]: integrations

**Comment by jwhited** [comment]
Do we have an associated Tailscale commit where the interface is implemented?

Any idea around impact to performance?

**Review by jwhited — APPROVED** [review]

**Comment by bradfitz** [comment]
I fixed the typo and verified this code path isn't hot (or even called) during normal operation, so I think this is fine to merge & we can experiment with (and test) in oss.

**Comment by bradfitz** [comment]
(merging per chat with @raggi the other day)

**Pull request merged**